tree/ntuple/v7/src/RNTupleProcessor.cxx

```diff
@@ -352,135 +349,122 @@ ROOT::Experimental::RNTupleJoinProcessor::RNTupleJoinProcessor(const RNTupleOpen
    }
 }
 
-void ROOT::Experimental::RNTupleJoinProcessor::AddAuxiliary(const RNTupleOpenSpec &auxNTuple,
-                                                            const std::vector<std::string> &joinFields,
-                                                            std::unique_ptr<RNTupleModel> model)
+void ROOT::Experimental::RNTupleJoinProcessor::SetModel(std::unique_ptr<RNTupleModel> primaryModel,
+                                                        std::vector<std::unique_ptr<RNTupleModel>> auxModels)
 {
-   assert(fNEntriesProcessed == 0 && "cannot add auxiliary ntuples after processing has started");
-
-   fNTuples.emplace_back(auxNTuple);
-
-   auto pageSource = Internal::RPageSource::Create(auxNTuple.fNTupleName, auxNTuple.fStorage);
-   pageSource->Attach();
-
-   if (pageSource->GetNEntries() == 0) {
-      throw RException(R__FAIL("provided RNTuple is empty"));
-   }
-
-   if (!model)
-      model = pageSource->GetSharedDescriptorGuard()->CreateModel();
+   fModel = std::move(primaryModel);
+   fModel->Unfreeze();
 
-   model->Freeze();
-   auto entry = model->CreateBareEntry();
+   // Create an anonymous record field for each auxiliary ntuple, containing their top-level fields. These original
+   // top-level fields are registered as subfields in the join model, such that they can be accessed as
+   // `auxNTupleName.fieldName`.
+   // TODO(fdegeus) Support projected fields
+   for (unsigned i = 0; i < auxModels.size(); ++i) {
+      std::vector<std::unique_ptr<RFieldBase>> auxFields;
 
-   // Append the auxiliary fields to the join model
-   fModel->Unfreeze();
+      if (!auxModels[i]) {
+         fAuxiliaryPageSources[i]->Attach();
+         auxModels[i] = fAuxiliaryPageSources[i]->GetSharedDescriptorGuard()->CreateModel();
+      }
 
-   // The fields of the auxiliary ntuple are contained in an anonymous record field and subsequently registered as
-   // subfields to the join model. This way they can be accessed through the processor as `auxNTupleName.fieldName`,
-   // which is necessary in case there are duplicate field names between the main ntuple and (any of the) auxiliary
-   // ntuples.
-   std::vector<std::unique_ptr<RFieldBase>> auxFields;
-   auxFields.reserve(entry->fValues.size());
-   for (const auto &val : *entry) {
-      auto &field = val.GetField();
+      for (const auto &fieldName : auxModels[i]->GetFieldNames()) {
+         auxFields.emplace_back(auxModels[i]->GetConstField(fieldName).Clone(fieldName));
+      }
 
-      auxFields.emplace_back(field.Clone(field.GetQualifiedFieldName()));
-   }
-   std::unique_ptr<RFieldBase> auxParentField =
-      std::make_unique<RRecordField>(auxNTuple.fNTupleName, std::move(auxFields));
+      auto auxParentField = std::make_unique<RRecordField>(fNTuples[i + 1].fNTupleName, std::move(auxFields));
+      if (!auxParentField) {
+         throw RException(R__FAIL("cannot add fields of auxiliary RNTuple \"" + fNTuples[i + 1].fNTupleName + "\""));
+      }
```

Comment:
Under which conditions can this happen, other than out-of-memory?

Reply:
None that I can think of right now. I've wrapped it in an `R__unlikely` now, but I'm also fine with removing it altogether.

Reply:
I believe if we don't check other return values of `std::make_unique` it should be removed here. If the constructor of `RRecordField` throws, the exception will be propagated on its own.